include/ur_api.h

```diff
@@ -2739,6 +2765,41 @@ typedef enum ur_usm_alloc_info_t
 
 } ur_usm_alloc_info_t;
 
+///////////////////////////////////////////////////////////////////////////////
+/// @brief USM allocation descriptor type
+typedef struct ur_usm_desc_t
+{
+    ur_structure_type_t stype;                      ///< [in] type of this structure
+    const void* pNext;                              ///< [in][optional] pointer to extension-specific structure
+    ur_usm_mem_flags_t flags;                       ///< [in] memory allocation flags
+    ur_mem_advice_t hints;                          ///< [in] Memory advice hints
```

Comment:
There's a mismatch between the granularity of allocations and the regions (pages) of memory to which these hints can actually apply.
Are implementations expected to separate allocations with different hints? Is it legal to allocate from the same pool using different hints? Are implementations again expected to create internal 'mini-pools'?

I would prefer to move hints to `ur_usm_pool_desc_t`.

Reply:
The idea I had in mind was to allow UR to decide how to handle each allocation. If the ur_usm_desc_t::pool is NULL, then UR can either request memory directly from the driver or create/use some internal memory pool. If the ur_usm_desc_t::pool is not NULL, UR could either ignore the hint or create a 'sub-pool' with desired `mem_advice` flag and allocate from that.

The reason I don't like putting `ur_mem_advice_t hints` (and `ur_usm_mem_flags_t flags` as well) is that the user would end up with a huge number of pools if he wants to support `ur_mem_advice_t` (we might have a pool pre-context, per-device, per-mem_advice, ...)

Reply:
I don't think we can leave this entirely up to implementations. At the very least, UR needs to define that hints can't affect other unrelated allocations.

I don't have a better idea. I agree that my proposal might not be super ergonomic, but what this is doing is mixing object management and page management in a single function, which is traditionally separate. Anyway, I don't feel super strongly about this, so feel free to ignore this suggestion.

Reply:
Yes, I definitely agree that we need to define that flags/hints do not affect unrelated allocations.

The long-term goal is still to expose lower-level API (through UMA and UR) that would allow you to create arbitrary pools. Given that, I think it makes sense to have a more high-level allocation API for now (leaving more to the UR).

I am also wondering if it wouldn't be better to rename 'pool' to something else (for the API proposed by this PR). Once we have a UMA integration in the future, it might be confusing to have two different 'pool' entities. But I don't really have a better name for now.